### PR TITLE
8257436: [aarch64] Regressions in ArrayCopyUnalignedDst.testByte/testChar for 65-78 bytes when UseSIMDForMemoryOps is on

### PR DESCRIPTION
**Pull request by eastig**
Copying data in range 65 - 80 bytes with
```
ldpq
ldpq 
ldp
stpq
stpq
stp
```
is faster than with
```
ldpq
ldpq
ldpq
stpq
stpq
stpq
```
when pointers involved are not properly aligned. This mostly affects byte and char arraycopy. 
This patch fixes regressions of ArrayCopyUnalignedDst.testByte/testChar when UseSIMDForMemoryOps is on.
See the attached results for performance data
[test_byte_results.txt](https://github.com/openjdk/jdk/files/5629676/test_byte_results.txt)
[test_char_results.txt](https://github.com/openjdk/jdk/files/5629673/test_char_results.txt)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257436](https://bugs.openjdk.java.net/browse/JDK-8257436): [aarch64] Regressions in ArrayCopyUnalignedDst.testByte/testChar for 65-78 bytes when UseSIMDForMemoryOps is on

### Reviewers
 * [Volker Simonis](https://openjdk.java.net/census#simonis) (@simonis - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1558/head:pull/1558`
`$ git checkout pull/1558`

### CODE REVIEW

**Comment by eastig** [comment]
@theRealAph could please check if it causes any issues on ThunderX 2?

Thanks,
Evgeny

**Comment by theRealAph** [comment]
> @theRealAph could please check if it causes any issues on ThunderX 2?

Is there some script you use that generates the comparison? Otherwise it'll take some time.

**Comment by eastig** [comment]
> > @theRealAph could please check if it causes any issues on ThunderX 2?
> 
> Is there some script you use that generates the comparison? Otherwise it'll take some time.

Unfortunately, no. I have scripts that do runs. Then I pull all results in Excel.
BTW, I attached data comparing SIMD_OFF vs SIMD_ON(patch). JDK-8257436 has data showing regressions with SIMD three ldpq instructions. I'll attach data comparing SIMD_no_fix vs SIMD_fix later.

**Comment by eastig** [comment]
I updated the attached data. It now includes the following:
- UseSIMDForMemoryOps is off.
- UseSIMDForMemoryOps is on.
- UseSIMDForMemoryOps + fix is on.

The fix is narrowed to be used for byte and char arraycopies. They have the most problems with unaligned accesses. Int arraycopies are at least 4 byte aligned. Long arraycopies are always 8 byte aligned.

**Comment by eastig** [comment]
I am attaching performance data for int arraycopy to show the impact of the patch on it:
[test_int_results.txt](https://github.com/openjdk/jdk/files/5631318/test_int_results.txt)

**Review by simonis — CHANGES_REQUESTED** [review]
Hi Evegeny,

thanks for fixing this. Your change looks good to me. Thanks especially for measuring all the different cases.

I only have two minor request regarding the comments. Once fixed I'll sponsor this change.

Best regards,
Volker

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 1]
```diff
@@ -1154,7 +1154,28 @@ class StubGenerator: public StubCodeGenerator {
     if (UseSIMDForMemoryOps) {
       __ ldpq(v0, v1, Address(s, 0));
       __ ldpq(v2, v3, Address(s, 32));
+      // Unaligned pointers can be an issue for copying.
+      // This have more chances to happen when granularity of data is
```
[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp]: Done

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 2]
```diff
@@ -1094,7 +1094,7 @@ class StubGenerator: public StubCodeGenerator {
                    Register count, Register tmp, int step) {
     copy_direction direction = step < 0 ? copy_backwards : copy_forwards;
     bool is_backwards = step < 0;
-    int granularity = uabs(step);
+    unsigned int granularity = uabs(step);
     const Register t0 = r3, t1 = r4;
 
     // <= 96 bytes do inline. Direction doesn't matter because we always
```
[the remark this answers is not in the document]
Comment (eastig) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp]: Done.

**Review by simonis — APPROVED** [review]
Looks good now. Thanks!

**Pull request closed without merging**